dpnp/dpnp_utils/dpnp_utils_statistics.py

```diff
@@ -23,11 +23,108 @@
 # THE POSSIBILITY OF SUCH DAMAGE.
 # *****************************************************************************
 
+import warnings
+
+import dpctl
+import dpctl.tensor as dpt
+from dpctl.tensor._numpy_helper import normalize_axis_tuple
+from dpctl.utils import ExecutionPlacementError
 
 import dpnp
+from dpnp.dpnp_array import dpnp_array
 from dpnp.dpnp_utils import get_usm_allocations, map_dtype_to_device
 
-__all__ = ["dpnp_cov"]
+__all__ = ["dpnp_cov", "dpnp_median"]
+
+
+def _calc_median(a, axis, out=None):
+    """Compute the median of an array along a specified axis."""
+
+    indexer = [slice(None)] * a.ndim
+    index, remainder = divmod(a.shape[axis], 2)
+    if remainder == 1:
+        # index with slice to allow mean (below) to work
+        indexer[axis] = slice(index, index + 1)
+    else:
+        indexer[axis] = slice(index - 1, index + 1)
+
+    # Use `mean` in odd and even case to coerce data type and use `out` array
+    res = dpnp.mean(a[tuple(indexer)], axis=axis, out=out)
+    nan_mask = dpnp.isnan(a).any(axis=axis)
+    if nan_mask.any():
+        res[nan_mask] = dpnp.nan
+
+    return res
+
+
+def _calc_nanmedian(a, axis, out=None):
+    """Compute the median of an array along a specified axis, ignoring NaNs."""
+    mask = dpnp.isnan(a)
+    valid_counts = dpnp.sum(~mask, axis=axis)
+    if out is None:
+        res = dpnp.empty_like(valid_counts, dtype=a.dtype)
+    else:
+        dpnp.check_supported_arrays_type(out)
+        exec_q = dpctl.utils.get_execution_queue((a.sycl_queue, out.sycl_queue))
+        if exec_q is None:
+            raise ExecutionPlacementError(
+                "Input and output allocation queues are not compatible"
+            )
+        if out.shape != valid_counts.shape:
+            raise ValueError(
+                f"Output array of shape {valid_counts.shape} is needed, got {out.shape}."
+            )
+        res = out
+
+    # Iterate over all indices of the output shape
+    for idx in dpnp.ndindex(res.shape):
+        current_valid_counts = valid_counts[idx]
+
+        if current_valid_counts > 0:
```

Comment:
There is synchronization at many places inside the loop (any call of `__bool__()`).
I wonder if it will be faster comparing to numpy-like implementation:
```python
def _remove_nan_1d(arr1d, overwrite_input=False):
    c = dpnp.isnan(arr1d)

    s = dpnp.nonzero(c)[0]
    if s.size == arr1d.size:
        warnings.warn("All-NaN slice encountered", RuntimeWarning, stacklevel=6)
        return arr1d[:0], True
    elif s.size == 0:
        return arr1d, overwrite_input
    else:
        if not overwrite_input:
            arr1d = arr1d.copy()

        # select non-nans at end of array
        enonan = arr1d[-s.size:][~c[-s.size:]]
        # fill nans in beginning of array with non-nans of end
        arr1d[s[:enonan.size]] = enonan
        return arr1d[:-s.size], True


def _nanmedian1d(arr1d, overwrite_input=False):
    arr1d_parsed, overwrite_input = _remove_nan_1d(arr1d, overwrite_input=overwrite_input)
    if arr1d_parsed.size == 0:
        # Ensure that a nan-esque scalar of the appropriate type (and unit)
        # is returned for `complexfloating`
        return arr1d[-1]

    return dpnp.median(arr1d_parsed, overwrite_input=overwrite_input)


def _calc_nanmedian(a, axis, out=None, overwrite_input=False):
    if axis is None or a.ndim == 1:
        part = a.ravel()
        if out is None:
            return _nanmedian1d(part, overwrite_input)
        else:
            out[...] = _nanmedian1d(part, overwrite_input)
            return out
    else:
        result = dpnp.apply_along_axis(_nanmedian1d, axis, a, overwrite_input)
        if out is not None:
            out[...] = result
        return result
```
I think the main difference might be avoiding `__bool__()` calls and a special use case to handle 1D arrays.

Reply:
This alternative approach is implemented here: https://github.com/IntelPython/dpnp/compare/master...nanmedian-numpy

Timing does not show an improvement over the current implementation (compare with timing presented for current PR [here](https://github.com/IntelPython/dpnp/pull/2191#issuecomment-2515960303).

In the current PR, input array is first sorted along axis and `nan` are sent to the end of each axis and then median is calculated for each axis by looping over axes while ignoring `nan` and the end of each array.
However, in the alternative approach, first `nan` are handled along each axis and then sorting is also done along each axis and then the median is calculated. In other words, in the current PR, only calculating median is performed along each axis (sorting and counting `nan` is performed over the whole array at once) while in the alternative approach handling `nan`, sorting and finding median is completed along each axis separately.


| `nanmedian` Function      | NumPy                | CPU-Xeon             | GPU-PVC              |
|------------------------|----------------------|----------------------|----------------------|
| (2048, 2048), axis=None   | 18.9 ms ± 39.4 μs    | 29.8 ms ± 5.27 ms    | 3.5 ms ± 10.2 μs     |
| (2048, 2048), axis=0      | 74.7 ms ± 55.1 μs    | 4.42 s ± 69.6 ms     | 1.6 s ± 4.83 ms      |
| (2048, 2048), axis=1      | 35.3 ms ± 59.6 μs    | 4.01 s ± 159 ms      | 1.51 s ± 5.99 ms     |
|                        |                      |                      |                      |
| (16*8192, 32), axis=None  | 18.9 ms ± 34.4 μs    | 31.1 ms ± 8.15 ms    | 3.41 ms ± 4.6 μs     |
| (16*8192, 32), axis=0     | 52.1 ms ± 47.1 μs    | 183 ms ± 15.1 ms     | 41.6 ms ± 93.2 μs    |
| (16*8192, 32), axis=1     | 87.8 ms ± 157 μs     | 2min 56s ± 835 ms    | 1min 12s ± 373 ms    |